makemoduledefs: Avoid empty module lists.

jepler opened the pull request:
### Summary

An empty array is a C extension supported by clang & gcc but not msvc. If the extensible module list is empty, eliminate the global variable and all code related to it. This is expected to not only resolve the problem building micropython but also saves code in the minimal build, which has no extensible modules.

Closes: #18141

### Testing

I ran a unix "make test" locally. I also built the "minimal" port, which at least exercises the case of empty *extensible* module list. (saves 256 bytes of text locally)

### Trade-offs and Alternatives

In #18141, @Robert-m-lucas suggested always adding a NULL entry at the end of the table, but this is probably undesirable since it would require extra text everywhere.

Initially this PR also handled the case of an empty non-extensible module list but it appears at least one such module will always exist (`__main__` in runtime.c)

Closes: #18141
Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
makemoduledefs: Avoid empty extensible module lists.

An empty array is a C extension supported by clang & gcc
but not msvc.

This also saves a bit of code size if there are no extensible
modules.

Closes: #18141
Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
py: Regenerate moduledefs.h if makemoduledefs.py changes.

This makes iterating on the script easier.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

dpgeorge commented on py/objmodule.c:
```diff
@@ -173,7 +175,11 @@ static const mp_module_delegation_entry_t mp_builtin_module_delegation_table[] =
 // Attempts to find (and initialise) a built-in, otherwise returns
 // MP_OBJ_NULL.
 mp_obj_t mp_module_get_builtin(qstr module_name, bool extensible) {
```
Maybe this function should have signature ` mp_obj_t mp_module_get_builtin(qstr module_name)` when extensible modules are disabled?  That might reduce code size a tiny bit more, be more obvious what's happening, and prevent unintended use of the `extensible` argument.

dpgeorge commented on py/objmodule.c:
```diff
@@ -173,7 +175,11 @@ static const mp_module_delegation_entry_t mp_builtin_module_delegation_table[] =
     mp_map_elem_t *elem = mp_map_lookup((mp_map_t *)(extensible ? &mp_builtin_extensible_module_map : &mp_builtin_module_map), MP_OBJ_NEW_QSTR(module_name), MP_MAP_LOOKUP);
+    #else
+    mp_map_elem_t *elem = mp_map_lookup((mp_map_t *)&mp_builtin_module_map, MP_OBJ_NEW_QSTR(module_name), MP_MAP_LOOKUP);
+    #endif
```
Instead of this, suggest:
```c
#if MICROPY_HAVE_REGISTERED_EXTENSIBLE_MODULES
mp_map_t *map = extensible ? &mp_builtin_extensible_module_map : &mp_builtin_module_map;
#else
mp_map_t *map = &mp_builtin_module_map;
#endif
mp_map_elem_t *elem = mp_map_lookup(map, MP_OBJ_NEW_QSTR(module_name), MP_MAP_LOOKUP);
```

jepler replied:
done!

dpgeorge commented:
If you could please rebase on latest master, it'll pick up all the new unix minimal (and zephyr minimal) tests.

jepler committed:
objmodule: Eliminate "bool extensible" param when possible.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

dpgeorge commented on py/objmodule.h:
```diff
@@ -35,10 +35,20 @@
+#define mp_module_get_builtin(module_name, extensible) ( \
+    MP_STATIC_ASSERT(!extensible), mp_module_get_builtin_impl((module_name)) \
+    )
+#endif
```
This looks OK.  An alternative (not saying it would be better this way!) is:
```c
#if MICROPY_HAVE_REGISTERED_EXTENSIBLE_MODULES
mp_obj_t mp_module_get_builtin_impl(qstr module_name, bool extensible);
#define mp_module_get_builtin(module_name) mp_module_get_builtin_impl((module_name), false)
#define mp_module_get_builtin_extensible(module_name) mp_module_get_builtin_impl((module_name), true)
#else
mp_obj_t mp_module_get_builtin(qstr module_name);
#endif
```
then change all uses of `mp_module_get_builtin(module, false)` to `mp_module_get_builtin(module)` and change `mp_module_get_builtin(module, true)` to `mp_module_get_builtin_extensible(module)`.

jepler replied:
Implemented in its own commit, so you can take it or not at your preference.

dpgeorge commented on py/objmodule.h:
```diff
@@ -35,10 +35,20 @@
+extern const mp_map_t mp_builtin_extensible_module_map;
+#endif
 
+#if MICROPY_HAVE_REGISTERED_EXTENSIBLE_MODULES
```
Could probably combine this and the above `#if` into one `#if` block.

jepler replied:
done.

jepler committed:
objmodule: Split get_builtin/_extensible into 2 macros.

This makes the intent clearer at the call sites.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>

jepler committed:
objmodule: Avoid interning a string unnecessarily.

If the lookup of the module name less the leading "u"
is going to succeed, it's already a qstr. So, use a
function that will avoid interning it if it's not already.

e.g., if you `import unavailable` it used to intern the
string `navailable`, but now it won't.

Signed-off-by: Jeff Epler <jepler@unpythonic.net>
